megatron/optimizer/optimizer.py

```diff
@@ -272,23 +272,27 @@ def allreduce_key_value_grads(self, args):
         Coalesce the bias grads to avoid too many small reductions,
         but not the weight grads since it could cause memory issues.
         """
-        grads=[]
-        for model_module in self.models:
-            unwrapped_model = unwrap_model(
-                    model_module, (torchDDP, LocalDDP, Float16Module))
-            for layer in unwrapped_model.language_model.encoder.layers:
-                kv_weight = layer.self_attention.key_value.weight
-                grad = kv_weight.main_grad if args.DDP_impl == 'local' else kv_weight.grad
-                torch.distributed.all_reduce(grad, group=mpu.get_tensor_model_parallel_group())
-                kv_bias = layer.self_attention.key_value.bias
-                grads.append(kv_bias.main_grad if args.DDP_impl == 'local' else kv_bias.grad)
-        if len(grads)>0:
-            coalesced = _flatten_dense_tensors(grads)
-            torch.distributed.all_reduce(
-                coalesced, group=mpu.get_tensor_model_parallel_group())
-            for buf, synced in zip(grads, _unflatten_dense_tensors(
-                    coalesced, grads)):
-                buf.copy_(synced)
+        # All-reduce kv parameters across model parallel nodes
+        # when sequence parallelism is used
+        if mpu.get_tensor_model_parallel_world_size() > 1 and \
```

Comment:
Can you please move the conditions outside this function and put it with the checks for mqa? (https://github.com/bigcode-project/Megatron-LM/pull/37/files#diff-4125b0b128f265b3cdd9786f9419957a9bb4c75a93aba7e17fb0ba3fed1da199L545, https://github.com/bigcode-project/Megatron-LM/pull/37/files#diff-dabd5ffec3ce00acedf6bf253e236186da42c61e2501d7d3f5a84645bfb6578bL337)
This would reduce the diff (the rest of this function is unchanged right?) and skip some unnecessary synchronization with the timers.